### DISCUSSION TITLE
Migrating an existing repo to GIT LFS

### DISCUSSION THREAD

**Discussion by SalmaFarghaly** [git-lfs/git-lfs#5011] [Q&A] [answered]
Hello,
I have some questions about migrating an existing repo to GIT LFS

1) What can go wrong that if I decided to use LFS on an existing repo with only running git lfs track <pattern> and git add --renormalize .
and not using "git lfs migrate import" ? I know that migrate command rewrite the history of the repo, So what are the pros and cons of rewriting the history of repo for using lfs?

2)  When I run the migrate commands using "git lfs migrate import --everything --include="*.txt"; git push --all --force;" . I can see that content of text files has changed to something like that : `version https://git-lfs.github.com/spec/v1
oid sha256:9db7a3f9a2a036bce91807cc58fd8b90fec8ab2085f4760a9eeba97430b6aa79
size 11` and the old content has gone. How I can get the old content of file? 

3) If I decided to migrate existing repo to GIT LFS using "git lfs migrate import --everything --include="*.txt"; git push --all --force;" . What should other clients to the repo do to get the migration effect ? Should they run "git pull" command only ?

**Comment by bk2204** [comment, marked as answer]
Hey,

For case 1, it's fine to migrate only the current revision that way.  However, if you have large files in the history, they won't be removed, so your repository may remain larger than it would be if you hadn't removed them.  If that's not a problem, this is a perfectly fine way to go about migrating your repository, and since it also avoids needing to rewrite history, you don't have to disrupt current users of the repository.

In question 2, those are pointer files, and they replace the large files.  If you've run `git lfs install` and you've pushed all the LFS files (which you can do with `git lfs push --all`), then normally when you check out the repository, the LFS files will be downloaded and checked out automatically by Git LFS.  If you've pushed the files but haven't run `git lfs install`, then you can check them out with `git lfs pull`.  Note, however, that commands like `git show` and `git diff` typically still show the pointer file.  You can ask Git to show LFS files in such a case [by using the technique described in the FAQ](https://github.com/git-lfs/git-lfs/wiki/FAQ).

If you rewrite history in question 3, then `git pull` won't work, because you'll have divergent histories.  Even if you somehow don't have divergent histories, you won't want to use `git pull` because it will try to do a giant merge of the histories which will fail with large numbers of conflicts, and even if those are resolved, the result will not be what you want.  You'll want to have users instead reset their branches with something like `git reset --hard origin/main` (after verifying they have no local changes).  They could then rebase their feature branches with something like `git rebase --onto origin/main main`.

**Reply by SalmaFarghaly** [reply]
Hey @bk2204 
Thank you for your response . If I moved on with "git lfs track and git add --renormalize ." solution . Would it be okay If I decided to run the migration commands later on after already the "git lfs track and git add --renormalize ." has been applied to repo ?

**Reply by bk2204** [reply]
Yes, that should work fine.

**Reply by SalmaFarghaly** [reply]
@bk2204 
Is it okay If every user on repo run "git lfs track and git add --renormalize ." on each branch ? . I have tried the following to check that.
Me and my colleague have "main" branch that has temp.txt file , and none of us has local changes to that file. and we did the following 
Me : While main branch is checked out , I tried to run "git lfs track "*.txt" ; git lfs track "*.txt" --lockable ; git add .gitattributes ; git add --renormalize .  ; git commit --m "Applying locking mechanism"; GIT_LFS_SKIP_SMUDGE=1 ; git read-tree -u --reset HEAD ;" 
My colleague : he ran the same command on "main" branch too. Then he pushed his changes " git push origin main". 
Me : tried to pull the "main" But I got merge conflict similar to that in temp.txt
>version https://git-lfs.github.com/spec/v1
<<HEAD
oid sha256:f0b99dae270925c18017d29308b5bf29bba7f20d77c71475d6ec919bd7c335dd
size 12
=====
oid sha256:9db7a3f9a2a036bce91807cc58fd8b90fec8ab2085f4760a9eeba97430b6aa79
size 11
>5aec96ccefdc80da7307b77b7e22134ada339c88

I want to conclude that to apply lfs correctly on existing repo without using migration . ONLY one user should run "git lfs track and git add --renormalize ." on each branch given that user has pulled all latest changes before running command and he has no local changes . Is that conclusion right ?

**Reply by bk2204** [reply]
Yes, you do.  That's because otherwise you'll run into conflicts when you try to merge and one side has an LFS file and the other has a plain Git file.
